File: JavaScript/WwiseSoundbankParser.js
// ©2023 JDSherbert

function parseWwiseSoundbankXML(soundbankXml) 
{
  // Parse the XML
  const parser = new DOMParser();
  const xmlDoc = parser.parseFromString(soundbankXml, "text/xml");

  // Initialize the result objects
  const eventMap = new Map();
  const rtpcMap = new Map();
  const switchMap = new Map();
  const stateMap = new Map();
  const auxBusMap = new Map();

  // Retrieve event IDs and names
  const events = xmlDoc.getElementsByTagName("Event");
  for (const event of events) {
    const eventID = event.getElementsByTagName("ID")[0].textContent;
    const eventName = event.getElementsByTagName("Name")[0].textContent;
    eventMap.set(eventID, eventName);
  }

  // Retrieve RTPC IDs and names
  const rtpcs = xmlDoc.getElementsByTagName("RealTimeParameter");
  for (const rtpc of rtpcs) 
  {
    const rtpcID = rtpc.getElementsByTagName("ID")[0].textContent;
    const rtpcName = rtpc.getElementsByTagName("Name")[0].textContent;
    rtpcMap.set(rtpcID, rtpcName);
  }

  // Retrieve switch IDs and names
  const switches = xmlDoc.getElementsByTagName("Switch");
  for (const switchElem of switches) 
  {
    const switchID = switchElem.getElementsByTagName("ID")[0].textContent;
    const switchName = switchElem.getElementsByTagName("Name")[0].textContent;
    switchMap.set(switchID, switchName);
  }

  // Retrieve state IDs and names
  const states = xmlDoc.getElementsByTagName("State");
  for (const state of states) 
  {
    const stateID = state.getElementsByTagName("ID")[0].textContent;
    const stateName = state.getElementsByTagName("Name")[0].textContent;
    stateMap.set(stateID, stateName);
  }

  // Retrieve aux bus IDs and names
  const auxBuses = xmlDoc.getElementsByTagName("AuxBus");
  for (const auxBus of auxBuses) 
  {
    const auxBusID = auxBus.getElementsByTagName("ID")[0].textContent;
    const auxBusName = auxBus.getElementsByTagName("Name")[0].textContent;
    auxBusMap.set(auxBusID, auxBusName);
  }

  // Return the result objects
  return 
  {
    eventMap,
    rtpcMap,
    switchMap,
    stateMap,
    auxBusMap,
  };
}

File: C#/WwiseSoundbankParser.cs
// ©2023 JDSherbert

using System.Collections.Generic;
using System.Xml;

public static class WwiseSoundbankParser
{
    public static Dictionary<string, string> GetEventIDAndName(string soundbankXmlPath)
    {
        XmlDocument doc = new XmlDocument();
        doc.Load(soundbankXmlPath);

        XmlNodeList eventNodes = doc.SelectNodes("/SoundBank/Events/Event");

        Dictionary<string, string> eventDict = new Dictionary<string, string>();

        foreach (XmlNode eventNode in eventNodes)
        {
            string eventID = eventNode.SelectSingleNode("ID")?.InnerText;
            string eventName = eventNode.SelectSingleNode("Name")?.InnerText;

            if (!string.IsNullOrEmpty(eventID) && !string.IsNullOrEmpty(eventName))
            {
                eventDict[eventID] = eventName;
            }
        }

        return eventDict;
    }

    public static Dictionary<string, string> GetRTPCIDAndName(string soundbankXmlPath)
    {
        XmlDocument doc = new XmlDocument();
        doc.Load(soundbankXmlPath);

        XmlNodeList rtpcNodes = doc.SelectNodes("/SoundBank/RealTimeParameters/RealTimeParameter");

        Dictionary<string, string> rtpcDict = new Dictionary<string, string>();

        foreach (XmlNode rtpcNode in rtpcNodes)
        {
            string rtpcID = rtpcNode.SelectSingleNode("ID")?.InnerText;
            string rtpcName = rtpcNode.SelectSingleNode("Name")?.InnerText;

            if (!string.IsNullOrEmpty(rtpcID) && !string.IsNullOrEmpty(rtpcName))
            {
                rtpcDict[rtpcID] = rtpcName;
            }
        }

        return rtpcDict;
    }

    public static Dictionary<string, string> GetSwitchIDAndName(string soundbankXmlPath)
    {
        XmlDocument doc = new XmlDocument();
        doc.Load(soundbankXmlPath);

        XmlNodeList switchNodes = doc.SelectNodes("/SoundBank/Switches/Switch");

        Dictionary<string, string> switchDict = new Dictionary<string, string>();

        foreach (XmlNode switchNode in switchNodes)
        {
            string switchID = switchNode.SelectSingleNode("ID")?.InnerText;
            string switchName = switchNode.SelectSingleNode("Name")?.InnerText;

            if (!string.IsNullOrEmpty(switchID) && !string.IsNullOrEmpty(switchName))
            {
                switchDict[switchID] = switchName;
            }
        }

        return switchDict;
    }

    public static Dictionary<string, string> GetStateIDAndName(string soundbankXmlPath)
    {
        XmlDocument doc = new XmlDocument();
        doc.Load(soundbankXmlPath);

        XmlNodeList stateNodes = doc.SelectNodes("/SoundBank/States/State");

        Dictionary<string, string> stateDict = new Dictionary<string, string>();

        foreach (XmlNode stateNode in stateNodes)
        {
            string stateID = stateNode.SelectSingleNode("ID")?.InnerText;
            string stateName = stateNode.SelectSingleNode("Name")?.InnerText;

            if (!string.IsNullOrEmpty(stateID) && !string.IsNullOrEmpty(stateName))
            {
                stateDict[stateID] = stateName;
            }
        }

        return stateDict;
    }

    public static Dictionary<string, string> GetAuxBusIDAndName(string soundbankXmlPath)
    {
        XmlDocument doc = new XmlDocument();
        doc.Load(soundbankXmlPath);

        XmlNodeList auxBusNodes = doc.SelectNodes("/SoundBank/AuxBus/AuxBus");

        Dictionary<string, string> auxBusDict = new Dictionary<string, string>();

        foreach (XmlNode auxBusNode in auxBusNodes)
        {
            string auxBusID = auxBusNode.SelectSingleNode("ID")?.InnerText;
            string auxBusName = auxBusNode.SelectSingleNode("Name")?.InnerText;

            if (!string.IsNullOrEmpty(auxBusID) && !string.IsNullOrEmpty(auxBusName))
            {
                auxBusDict[auxBusID] = auxBusName;
            }
        }

        return auxBusDict;
    }
}
